src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1384,37 +1385,83 @@ void FileCache::applySettingsIfPossible(const FileCacheSettings & new_settings,
+        /// b. Release a cache lock.
+        ///     1. Do actual eviction from filesystem.
         {
-            actual_settings.max_size = main_priority->getSizeLimit(cache_lock);
-            actual_settings.max_elements = main_priority->getElementsLimit(cache_lock);
-            throw;
+            cache_is_being_resized.store(true, std::memory_order_relaxed);
+            SCOPE_EXIT({
+                cache_is_being_resized.store(false, std::memory_order_relaxed);
+            });
+
+            auto cache_lock = lockCache();
+
+            FileCacheReserveStat stat;
+            if (main_priority->collectCandidatesForEviction(
+                    new_settings.max_size, new_settings.max_elements, 0/* max_candidates_to_evict */,
+                    stat, eviction_candidates, cache_lock))
+            {
+                /// Remove only queue entries of eviction candidates.
+                eviction_candidates.removeQueueEntries(cache_lock);
+                /// Note that (in-memory) metadata about corresponding file segments
+                /// (e.g. file segment info in CacheMetadata) will be removed
+                /// only after eviction from filesystem. This is needed to avoid
+                /// a race on removal of file from filesystsem and
+                /// addition of the same file as part of a newly cached file segment.
+
+                /// Modify cache size limits.
+                /// From this point cache eviction will follow them.
+                main_priority->modifySizeLimits(
+                    new_settings.max_size, new_settings.max_elements,
+                    new_settings.slru_size_ratio, cache_lock);
+
+                modified_size_limit = true;
+            }
         }
 
-        if (updated)
+        if (modified_size_limit)
         {
+            try
+            {
+                /// Do actual eviction from filesystem.
+                eviction_candidates.evict();
+            }
+            catch (...)
+            {
+                if (eviction_candidates.needFinalize())
+                    eviction_candidates.finalize(nullptr, lockCache());
+                throw;
+            }
+
+            if (eviction_candidates.needFinalize())
+                eviction_candidates.finalize(nullptr, lockCache());
+
             LOG_INFO(log, "Changed max_size from {} to {}, max_elements from {} to {}",
                     actual_settings.max_size, new_settings.max_size,
                     actual_settings.max_elements, new_settings.max_elements);
 
-            actual_settings.max_size = main_priority->getSizeLimit(cache_lock);
-            actual_settings.max_elements = main_priority->getElementsLimit(cache_lock);
+            actual_settings.max_size = new_settings.max_size;
+            actual_settings.max_elements = new_settings.max_elements;
+        }
+        else
+        {
+            LOG_WARNING(log, "Unable to modify size limit from {} to {}, "
```

Comment:
Will we try later, or only one attempt is made?

Reply:
This method is called for config reload, so no, another attempt will not be made until the next config reload.

Reply:
For better usability, we might even add a message to `system.warnings` like 'Your cache configuration is inconsistent with config.xml; try SYSTEM RELOAD CONFIG to make an attempt to shrink it.'
But it would most likely be overkill and not worth the bother.